### PR TITLE
Sweeping line triangulation 

### PR DESCRIPTION
**Pull request by Czaki**
This is the main algorithm for polygon triangulation. It requires that there is no intersecting edges in input.

### CODE REVIEW

**Comment by Czaki** [comment]
@coderabbitai review

File: crates/bermuda/src/lib.rs [hunk 1]
```diff
@@ -58,35 +68,97 @@ fn triangulate_path_edge(
         limit.unwrap_or(3.0),
         bevel.unwrap_or(false),
     );
-    let triangle_data: Vec<u32> = result
+    triangulation_to_edge_triangulate(py, &result)
+}
+
+fn triangulation_to_edge_triangulate(
```
Comment (willingc) [inline crates/bermuda/src/lib.rs:74]: Would it be more accurate to name `triangulation_to_edge_triangulation` since this seems to convert the data into a suitable python format?

Comment (willingc) [reply crates/bermuda/src/lib.rs:74]: Perhaps add a 1 line function docstring `Convert triangulation data to a Python edge triangulation`

File: crates/bermuda/src/lib.rs [hunk 2]
```diff
@@ -58,35 +68,97 @@ fn triangulate_path_edge(
...
             .into(),
         PyArray::from_vec(py, flat_offsets)
-            .reshape([result.offsets.len(), 2])?
+            .reshape([data.offsets.len(), 2])?
             .into(),
         triangle_array.into(),
     ))
 }
 
+fn triangulation_to_face_triangulate(
```
Comment (willingc) [inline crates/bermuda/src/lib.rs:104]:
```suggestion
fn triangulation_to_face_triangulation(
```

Same rationale as above with additional of similar function docstring.

File: crates/bermuda/src/lib.rs [hunk 3]
```diff
@@ -58,35 +68,97 @@ fn triangulate_path_edge(
...
+    ))
+}
+
+#[pyfunction]
+#[pyo3(signature = (polygons))]
+fn triangulate_polygons_with_edge(
+    py: Python<'_>,
+    polygons: Vec<PyReadonlyArray2<'_, f32>>,
+) -> PyPolygonTriangulation {
+    // Convert the numpy array into a rust compatible representations which is a vector of points.
```
Comment (willingc) [inline crates/bermuda/src/lib.rs:134]:
```suggestion
    // Convert the numpy array into a rust compatible representation which is a vector of points.
```

File: crates/bermuda/src/lib.rs [hunk 4]
```diff
@@ -58,35 +68,97 @@ fn triangulate_path_edge(
...
+        triangle_array.into(),
+        PyArray::from_vec(py, flat_points)
+            .reshape([points.len(), 2])?
+            .into(),
+    ))
+}
+
+#[pyfunction]
+#[pyo3(signature = (polygons))]
+fn triangulate_polygons_with_edge(
```
Comment (willingc) [inline crates/bermuda/src/lib.rs:130]: Adding a 1 line function docstring would be helpful to explain the function's purpose. Is the function perfoming: Convert polygons as vectors to a python polygon triangulation form.

File: crates/triangulation/src/path_triangulation.rs [hunk 5]
```diff
@@ -220,3 +220,28 @@ pub fn triangulate_path_edge(
     result.fix_triangle_orientation();
     result
 }
+
+pub fn triangulate_paths_edge(
```
Comment (willingc) [inline crates/triangulation/src/path_triangulation.rs:224]: Function name is a bit unclear. What does it mean to triangulate a path's edge? Is this "Convert paths represented by vectors to a Path Triangulation form."?

File: crates/triangulation/src/point.rs [hunk 6]
```diff
@@ -305,6 +306,14 @@ impl Triangle {
     pub fn new(x: Index, y: Index, z: Index) -> Self {
         Triangle { x, y, z }
     }
+
+    pub fn shifted_by(&self, shift: Index) -> Self {
```
Comment (willingc) [inline crates/triangulation/src/point.rs:310]: Instead of shift_by perhaps simply `shift`

File: tests/test_triangulate.py [hunk 7]
```diff
@@ -197,3 +197,238 @@ def test_change_limit():
...
+            [(2, 1, 0), (2, 0, 5), (4, 3, 2), (5, 4, 2)],
+        ),
+        ([(0, 0), (1, 2), (2, 0), (1, 1)], [(2, 1, 3), (3, 1, 0)]),
+        ([(0, 0), (0, 1), (0.5, 0.5), (1, 0), (1, 1)], [(3, 4, 2), (2, 1, 0)]),
+        ([(0, 0), (1, 0), (0.5, 0.5), (0, 1), (1, 1)], [(2, 4, 3), (1, 2, 0)]),
+    ]
+]
+
+
+def _renumerate_triangles(polygon, points, triangles):
```
Comment (willingc) [inline tests/test_triangulate.py:225]: Renumerate? What does that mean?

File: tests/test_triangulate.py [hunk 8]
```diff
@@ -197,3 +197,238 @@ def test_change_limit():
...
+            (4, 4),
+        ],
+        dtype=np.float32,
+    )
+    (triangles, points), _edges = triangulate_polygons_with_edge([polygon])
+    assert len(points) == 8
+    assert len(triangles) == 8
+
+
+def test_triangulate_polygon_segfault1():
```
Comment (willingc) [inline tests/test_triangulate.py:271]:
```suggestion
def test_triangulate_polygon_causes_segfault1():
```

Comment (willingc) [reply tests/test_triangulate.py:271]: Not sure why these tests are numbered 1, 2, 3, 4, 5 Could we parameterize these test cases by setting the polygons to constants and then run them through 1 test?

**Review by willingc — COMMENTED** [review]
Overall, this looks good. I do think it could be even better by being more explicit with some of names of functions and types so they are more clear to other devs. :sunny:

File: crates/bermuda/src/lib.rs (hunk 4 above)
Comment (willingc) [reply crates/bermuda/src/lib.rs:130]: Also not sure what the purpose of edges is?

File: crates/triangulation/src/point.rs (hunk 6 above)
Comment (Czaki) [reply crates/triangulation/src/point.rs:310]: For me, shift suggests doing it in place.

Comment (willingc) [reply crates/triangulation/src/point.rs:310]: Since most folks will be coming from a Python background where a Builder pattern is less explicit than Rust, I think the docstring will reduce the assumption that one is shifting in place.

File: tests/test_triangulate.py (hunk 7 above)
Comment (Czaki) [reply tests/test_triangulate.py:225]: Added docstring. Could you check now?

File: crates/triangulation/src/point.rs (hunk 6 above)
Comment (Czaki) [reply crates/triangulation/src/point.rs:311]: added docstring

File: crates/triangulation/src/path_triangulation.rs (hunk 5 above)
Comment (Czaki) [reply crates/triangulation/src/path_triangulation.rs:226]: Added docstring. Will think about better name.

File: crates/bermuda/src/lib.rs (hunk 4 above)
Comment (Czaki) [reply crates/bermuda/src/lib.rs:173]: Added docstring. Could you check?

File: crates/bermuda/src/lib.rs (hunk 2 above)
Comment (Czaki) [reply crates/bermuda/src/lib.rs]: changed name,

File: crates/bermuda/src/lib.rs (hunk 1 above)
Comment (Czaki) [reply crates/bermuda/src/lib.rs]: changed name

File: crates/triangulation/src/path_triangulation.rs (hunk 5 above)
Comment (willingc) [reply crates/triangulation/src/path_triangulation.rs:226]: I think the name is fine with the docstring. 👍🏼

**Review by willingc — APPROVED** [review]
Nice additions of docstrings. I would have been happy with 1 line but the additional context for the longer docstrings on a few of these is very helpful. Let's merge and release this package. :tada: Nice work.

File: tests/test_triangulate.py (hunk 8 above)
Comment (Czaki) [reply tests/test_triangulate.py:288]: These are five polygons that cause segfault when testing in C++ implementation. These are in separate functions as it simplify debugging. 
Also, in rust implementation they were useful to finding bugs in implementation.

Comment (willingc) [reply tests/test_triangulate.py:288]: Sounds good!

File: crates/bermuda/src/lib.rs (hunk 4 above)
Comment (willingc) [reply crates/bermuda/src/lib.rs:173]: Beautifully written and very understandable.

**Comment by Czaki** [comment]
I need to check a few polygons generated in [7632](https://github.com/napari/napari/pull/7632) for segfaults

File: tests/test_triangulate.py (hunk 7 above)
Comment (willingc) [reply tests/test_triangulate.py:225]: Nice!

**Review by willingc — APPROVED** [review]
:ship: Super helpful comments that you added. I would merge and release :tada:

**Comment by Czaki** [comment]
There are edge cases when both c++ and this implementation fails, but it may take longer to fix it so lets merge this and iterate in follow up.
